### PR TITLE
Dynamic pagination in OCP tab

### PR DESCRIPTION
**Pull request by MVarshini**
## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR includes handling Pagination, sorting and filtering of OCP page at the server side.

## Related Tickets & Documents

- Related Issue #
- Closes #

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please describe the System Under Test.
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: backend/app/api/v1/commons/ocp.py [hunk 1]
```diff
@@ -2,51 +2,196 @@
...
+def buildAggregateQuery():
+    keysDict = {"ciSystem":"ciSystem.keyword","platform":"platform.keyword","benchmark":"benchmark.keyword",
+                    "releaseStream":"releaseStream.keyword","networkType":"networkType.keyword", "workerNodesCount":"workerNodesCount","jobStatus":"jobStatus.keyword",
+                    "controlPlaneArch":"controlPlaneArch.keyword","publish":"publish.keyword","fips":"fips.keyword","encrypted":"encrypted.keyword",
+                    "ipsec":"ipsec.keyword", "ocpVersion":"ocpVersion.keyword", "build":"ocpVersion.keyword",
+                    "upstream":"upstreamJob.keyword"
+                }   
+    aggregate = {}
+    for x,y in keysDict.items():
+        obj = {x:{"terms":{"field":y,"size":10}}}
```
Comment (vishnuchalla) [inline backend/app/api/v1/commons/ocp.py:51]: Why do we have size hardcoded to 10 here?

File: backend/app/api/v1/commons/ocp.py [hunk 2]
```diff
@@ -2,51 +2,196 @@
...
+            query.update({"sort": json.loads(sort)})
+       
+        if bool(filter):
+            query = buildFilterQuery(filter, query) 
+            
+        aggregate = buildAggregateQuery()
+        query["aggs"].update(aggregate)
+        
+        isFilterReset = False
+        response = await es.filterPost(query=query)
```
Comment (vishnuchalla) [inline backend/app/api/v1/commons/ocp.py:139]: @MVarshini Did you make sure that this query is working?
```
RequestError was raised19: RequestError(400, 'search_phase_execution_exception', 'No mapping found for [build.keyword] in order to sort on')
TypeError was raised14: 'NoneType' object is not subscriptable
TypeError was raised15: argument of type 'NoneType' is not iterable
INFO:     127.0.0.1:42960 - "GET /api/v1/ocp/jobs?pretty=true&start_date=2024-10-10&end_date=2024-10-15&size=75&offset=1&sort=%7B%22build.keyword%22:%7B%22order%22:%22asc%22%7D%7D&filter=%7B%22ciSystem%22:[%22prow%22]%7D HTTP/1.1" 200 OK
Error retrieving filter data': RequestError(400, 'search_phase_execution_exception', 'No mapping found for [build.keyword] in order to sort on')
TypeError was raised: 'NoneType' object is not subscriptable
TypeError was raised: 'NoneType' object is not subscriptable
INFO:     127.0.0.1:42966 - "GET /api/v1/ocp/filters?pretty=true&start_date=2024-10-10&end_date=2024-10-15&size=75&offset=1&sort=%7B%22build.keyword%22:%7B%22order%22:%22asc%22%7D%7D&filter=%7B%22ciSystem%22:[%22prow%22]%7D HTTP/1.1" 200 OK
Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7feb9ce77be0>, 441476.621)]']
connector: <aiohttp.connector.TCPConnector object at 0x7feb8308b430>
{'query': {'query_string': {'query': 'uuid: "bc78c231-6d7f-4726-8afb-ef4a87d84331"'}}}
{'ciSystem': 'PROW', 'uuid': 'bc78c231-6d7f-4726-8afb-ef4a87d84331', 'releaseStream': '4.15.0-0.nightly', 'platform': 'AWS', 'clusterType': 'self-managed', 'benchmark': 'cluster-density-v2', 'masterNodesCount': 3, 'workerNodesCount': 6, 'infraNodesCount': 3, 'masterNodesType': 'm6a.xlarge', 'workerNodesType': 'm6a.xlarge', 'infraNodesType': 'r5.xlarge', 'totalNodesCount': 12, 'clusterName': 'ci-op-yfrkcx95-96149-rzgmn', 'ocpVersion': '4.15.0-0.nightly-2024-10-11-141906', 'networkType': 'OVNKubernetes', 'buildTag': '1844750966205714432', 'jobStatus': 'success', 'buildUrl': 'https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.15-nightly-x86-payload-control-plane-6nodes/1844750966205714432', 'upstreamJob': 'periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-aws-4.15-nightly-x86-payload-control-plane-6nodes', 'upstreamJobBuild': '4.15.0-0.nightly-2024-10-11-141906-qe-perfscale', 'executionDate': '2024-10-11T16:08:02Z', 'jobDuration': '1634', 'startDate': '2024-10-11T16:08:02Z', 'endDate': '2024-10-11T16:35:16Z', 'startDateUnixTimestamp': '1728662882', 'endDateUnixTimestamp': '1728664516', 'timestamp': '2024-10-11T16:35:29Z', 'ipsec': 'false', 'ipsecMode': 'Disabled', 'fips': 'false', 'encrypted': 'false', 'encryptionType': '', 'publish': 'External', 'computeArch': 'amd64', 'controlPlaneArch': 'amd64'}
{'query': {'bool': {'must': [{'query_string': {'query': 'benchmark: "cluster-density-v2$" AND workerNodesType: "m6a.xlarge" AND masterNodesType: "m6a.xlarge" AND masterNodesCount: "3" AND workerNodesCount: "6" AND platform: "AWS" AND ocpVersion: 4.15* AND jobStatus: success'}}]}}}
{'query': {'query_string': {'query': '( uuid: "bc78c231-6d7f-4726-8afb-ef4a87d84331" ) AND metricName: "jobSummary"'}}}
{'query': {'query_string': {'query': '( uuid: "f1055fa3-8c5c-4cc9-b75d-39594c499a58" OR uuid: "4d59851e-b665-4c6e-914d-fa7fea1e2cb0" OR uuid: "ce07640a-a094-411a-84d2-f24d5cd6d064" OR uuid: "14a0d6ae-d4e4-435e-95ab-2a37b5633d3c" OR uuid: "c2c59959-7980-4d04-8495-c42d75957e28" OR uuid: "7335a811-95a3-453b-ab8d-182b6c1d1c3e" OR uuid: "44ee7997-9327-4dff-8d42-e64a40110cfe" OR uuid: "28e7d49f-535e-43ca-a383-feff6e669908" OR uuid: "d80cfd21-d005-4085-8f4f-57117a9a016f" OR uuid: "d3b26ef5-54a5-44df-b025-e6465e00a596" ) AND metricName: "jobSummary"'}}}
INFO:     127.0.0.1:51932 - "GET /api/v1/ocp/graph/bc78c231-6d7f-4726-8afb-ef4a87d84331 HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/anyio/streams/memory.py", line 98, in receive
    return self.receive_nowait()
  File "/usr/local/lib/python3.9/site-packages/anyio/streams/memory.py", line 93, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

```
I see a bunch of errors in the dashboard backend pod. Also it doesn't render frontend UI at all while applying filters.

File: backend/app/services/search.py [hunk 3]
```diff
@@ -68,88 +55,115 @@ async def post(self, query, indice=None, size=10000, start_date=None, end_date=N
...
+                                size=size)
+                            previous_results = {"data":response['hits']['hits'], "total":response['hits']["total"]["value"]}
+                           # previous_results = await self.scan_indices(self.prev_es, self.prev_index, query, timestamp_field, start_date, new_end_date, size, offset)
                 if self.prev_es and self.new_es:
                     self.new_index = self.new_index_prefix + (self.new_index if indice is None else indice)
                     today = datetime.today().date()
                     seven_days_ago = today - timedelta(days=7)
                     if end_date and end_date < seven_days_ago:
-                        new_results = []
+                        new_results = {}
```
Comment (vishnuchalla) [inline backend/app/services/search.py:82]: Why is this being changed to a dictionary?

**Review by vishnuchalla — CHANGES_REQUESTED** [review]
Few questions in general while reviewing this PR
* How are the queries being changed to from the existing ones? What are the reasons behind making these choices?
* What is plan of implementation? .i.e in what order do we want our changes to be merged. Can we take them in small and atomic PRs breaking down into smaller components.
* How are we making sure that any of these changes are not breaking the existing components in the dashboard?


@MVarshini Can we please take a step back and come up with a design document with HLD and LLD for this work? I think we need an alignment on the execution/implementation plan first before we make any code changes. Thanks for understanding.

File: backend/app/api/v1/commons/ocp.py (hunk 2 above)
Comment (chentex) [reply backend/app/api/v1/commons/ocp.py:139]: Just minimize the search this is only happening to me when looking at `cluster-density-v2` jobs in ocp tab, seems to be realted to the Graph of this type of jobs

File: backend/app/services/search.py (hunk 3 above)
Comment (chentex) [reply backend/app/services/search.py:82]: Seems now it has extra values, I am guessing to deal with pagination, because now it includes the `total`

**Review by chentex — CHANGES_REQUESTED** [review]
I find that the `Quay` tab is not working.
Some comments regarding what Vishnu already pointed out.

**Comment by jtaleric** [comment]
Bad bot.

**Comment by jtaleric** [comment]
or is this a dupe @MVarshini ? @vishnuchalla maybe you know?

**Comment by vishnuchalla** [comment]
> or is this a dupe @MVarshini ? @vishnuchalla maybe you know?

Yeah, this one is duplicate. I will let @MVarshini to be kind to close this please in favor of https://github.com/cloud-bulldozer/cpt-dashboard/pull/135

**Comment by MVarshini** [comment]
Closing this PR in favour of PR[#135](https://github.com/cloud-bulldozer/cpt-dashboard/pull/135)

**Pull request closed without merging**